ompi/mpi/fortran/use-mpi-f08/cdesc/bsend_init_cdesc.c

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (c) 2004-2005 The Trustees of Indiana University and Indiana
+ *                         University Research and Technology
+ *                         Corporation.  All rights reserved.
+ * Copyright (c) 2004-2005 The University of Tennessee and The University
+ *                         of Tennessee Research Foundation.  All rights
+ *                         reserved.
+ * Copyright (c) 2004-2005 High Performance Computing Center Stuttgart,
+ *                         University of Stuttgart.  All rights reserved.
+ * Copyright (c) 2004-2005 The Regents of the University of California.
+ *                         All rights reserved.
+ * Copyright (c) 2011-2012 Cisco Systems, Inc.  All rights reserved.
+ * Copyright (c) 2015-2019 Research Organization for Information Science
+ *                         and Technology (RIST).  All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "ompi_config.h"
+
+#include "ompi/mpi/fortran/mpif-h/bindings.h"
+#include "ompi/mpi/fortran/use-mpi-f08/cdesc/bindings.h"
+#include "ompi/mpi/fortran/base/constants.h"
+
+
+void ompi_bsend_init_cdesc(CFI_cdesc_t* x, MPI_Fint *count, MPI_Fint *datatype, MPI_Fint *dest, MPI_Fint *tag, MPI_Fint *comm, MPI_Fint *request, MPI_Fint *ierr)
+{
+    int c_ierr;
+    MPI_Datatype c_datatype, c_type = PMPI_Type_f2c(*datatype);
+    MPI_Request c_req;
+    MPI_Comm c_comm;
+    void *buf = x->base_addr;
+    int c_count =  OMPI_FINT_2_INT(*count);
+
+    c_comm = PMPI_Comm_f2c (*comm);
+
+    c_datatype = c_type;
+    if (x->rank != 0 && !CFI_is_contiguous(x)) {
+        c_ierr = ompi_cdesc_create_datatype(x, c_count, c_type, &c_datatype);
+        if (MPI_SUCCESS != c_ierr) {
+            if (NULL != ierr) *ierr = OMPI_INT_2_FINT(c_ierr);
+            return;
+        }
+        c_count = 1;
+    }
```

Comment:
the code contains a `return` statement, so would you be fine using a macro instead?

Reply:
If it has to be a macro, that's probably still better than all the duplicated code.

...but just so that the question is asked 😉 -- is it possible to restructure a little bit to not require a `return` / allow it to be a proper function?